### ISSUE TITLE
Confusing edge case with @computed properties

### ISSUE THREAD

**Issue by butchler** [mobxjs/mobx#165]
I don't think this is a bug with Mobx, it's just a very confusing edge case
that I ran into while making a simple memory game (the one where you flip over
cards and try to remember where the card with the same image was) with Mobx and
React.

Here's a simplified version of the code in question:

```
import { observable, computed, autorun } from 'mobx';

class Card {
    constructor(game, id) {
        this.game = game;
        this.id = id;
    }

    @computed get isWrong() {
        return this.isSelected && this.game.isMatchWrong;
    }

    @computed get isSelected() {
        return this.game.firstCardSelected === this || this.game.secondCardSelected === this;
    }
}

class Game {
    @observable firstCardSelected = null;
    @observable secondCardSelected = null;

    @computed get isMatchWrong() {
        return this.secondCardSelected !== null && this.firstCardSelected.id !== this.secondCardSelected.id;
    }
}

let game = new Game();
let card1 = new Card(game, 1), card2 = new Card(game, 2);

autorun(() => {
    console.log('card1.isWrong =', card1.isWrong);
    console.log('card2.isWrong =', card2.isWrong);
});

// When the user selects the first card:
game.firstCardSelected = card1;
// Later, when the user selects the second card:
game.secondCardSelected = card2;
```

This outputs:

```
card1.isWrong = true
card2.isWrong = false
```

The second card doesn't consider itself to be false, even though the cards
don't match and they should both be wrong.

I added some logging to the computed properties to see what was going on:

```
import { observable, computed, autorun } from 'mobx';

class Card {
    constructor(game, id) {
        this.game = game;
        this.id = id;
    }

    @computed get isWrong() {
        console.log('Computing isWrong for card', this.id);
        return this.isSelected && this.game.isMatchWrong;
    }

    @computed get isSelected() {
        return this.game.firstCardSelected === this || this.game.secondCardSelected === this;
    }
}

class Game {
    @observable firstCardSelected = null;
    @observable secondCardSelected = null;

    @computed get isMatchWrong() {
        console.log('Computing isMatchWrong');
        return this.secondCardSelected !== null && this.firstCardSelected.id !== this.secondCardSelected.id;
    }
}

let game = new Game();
let card1 = new Card(game, 1), card2 = new Card(game, 2);

autorun(() => {
    console.log('card1.isWrong =', card1.isWrong);
    console.log('card2.isWrong =', card2.isWrong);
    console.log('------------------------------');
});

console.log('Selecting first card');
game.firstCardSelected = card1;
console.log('Selecting second card');
game.secondCardSelected = card2;
```

Which outputted:

```
Computing isWrong for card 1
card1.isWrong = false
Computing isWrong for card 2
card2.isWrong = false
------------------------------
Selecting first card
Computing isWrong for card 1
Computing isMatchWrong
Selecting second card
Computing isWrong for card 2
Computing isMatchWrong
Computing isWrong for card 1
card1.isWrong = true
card2.isWrong = false
------------------------------
```

After the second card is selected, it appears that card2.isWrong is computed
before isMatchWrong is recomputed, so it uses the old value of isMatchWrong,
leading to the incorrect answer. This is probably because Mobx doesn't know
that card2.isWrong depends on isMatchWrong because the && causes execution to
short-circuit when this.isSelected is false.

One way of fixing it is to just reverse the order of the condition:

```
    @computed get isWrong() {
        return this.game.isMatchWrong && this.isSelected;
    }
```

Or to explicitely tell Mobx that isWrong depends on isMatchWrong by just
accessing it in the getter:

```
    @computed get isWrong() {
        this.game.isMatchWrong;

        return this.isSelected && this.game.isMatchWrong;
    }
```

I might be doing something wrong, and I could probably rewrite the code in a
better way that would avoid this problem altogether. However, I just wanted to
let people know about this scenario in case anyone runs into a similar problem.

tl;dr: Be careful with short-circuiting caused by && or ||, because it seems
like it might cause Mobx to not be able to properly discover all of the
observables that a computed property depends on.

**Comment by butchler** [comment]
I am using Mobx version 2.0.2.

**Comment by mweststrate** [comment]
Thanks for reporting this one. While playing with your example, figured that in these cases the computed values should still be correct, even if it means computing the `isMatchWrong` function twice. I just released that as 2.0.3. But I will investigate whether the second (or first) computation could somehow be prevented.

**Comment by butchler** [comment]
Thank you for the really quick fix!
